### DISCUSSION TITLE
Query related to Training in Urdu and Arabic

### DISCUSSION THREAD

**Discussion by Anas-KhanWP** [SWivid/F5-TTS#620] [Q&A]
I have huge dataset for both Urdu and Arabic, around 1000 hours I am trying to train the model on both.

@SWivid, The only thing that I am concerned about is how will it differentiate between Urdu and Arabic? Both Vocabs are same but different pronunciation. How will the model distinguish between them?

@JarodMica, I am following your tutorial but you also chose to do 1 language only (Japanese) Can you help me out in this?

**Comment by SWivid** [comment]
@Anas-KhanWP hi, is the wording between them different? i mean, whether the organization of word is different though the characters are same.
if so, you can try bpe

**Reply by Anas-KhanWP** [reply]
@SWivid Let me show an example of both languages:

![image](https://github.com/user-attachments/assets/9ace1c5e-b9e5-41f7-9ec2-820d17338c9f)

Can you guide me now please?

**Reply by Anas-KhanWP** [reply]
I Tried using both pinyin and char tokenizer, but it is performing good on 1 language i.e Urdu.

**Reply by Anas-KhanWP** [reply]
Can you guide me through the process please?

**Reply by SWivid** [reply]
so the wording is different? i have not learn urdu or arabic so if you don't make it clear i just have no idea
maybe you can search for ipa or bpe tokenizers (thought some open-source repo there), and give them a try

**Reply by Anas-KhanWP** [reply]
Urdu language has every character that Arabic has but yes, Wording is different and Pronunciation as well. My current model is under training on Arabic and Urdu, it is just a low scale training just to test whether it can differentiate between both or not. So far all the generated samples are of Urdu languages, and they sound pretty decent. Correct me if I am wrong, but the samples would sound a bit of different of weird if it couldn't differentiate between them, right?

**Reply by SWivid** [reply]
The convergence speeds differentiate between languages, i.e., learned to speak one earlier, another later, it's acceptable.
Thought could train longer to see if works with char (it's important to be patient for a NAR model w/o phoneme force-alignment).
If after a certain training, the model mess up the nuance between languages, then you can turn to ipa or bpe.

**Reply by Anas-KhanWP** [reply]
Alright, Thanks @SWivid I will let you know how the model turns out.

**Reply by Anas-KhanWP** [reply]
@SWivid Need Assistance, I am trying to train it faster using multi gpu, As The previous model turned out to be really good. But I am getting this error:

GPU COUNT: 2
TOTAL MEMORY: 47.9755859375


accelerate launch --mixed_precision=bf16 src/f5_tts/train/finetune_cli.py --exp_name F5TTS_Base --learning_rate 1e-05 --batch_size_per_gpu 1600 --batch_size_type frame --max_samples 64 --grad_accumulation_steps 1 --max_grad_norm 1 --epochs 251 --num_warmup_updates 1904 --save_per_updates 3810 --last_per_steps 952 --dataset_name my_speak_dual_GPU --finetune True --tokenizer char  --log_samples True --logger tensorboard
W1221 03:52:52.024000 27048 torch\distributed\elastic\multiprocessing\redirects.py:28] NOTE: Redirects are currently not supported in Windows or MacOs.
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "E:\f5env\Scripts\accelerate.exe\__main__.py", line 7, in <module>
  File "E:\f5env\Lib\site-packages\accelerate\commands\accelerate_cli.py", line 48, in main
    args.func(args)
  File "E:\f5env\Lib\site-packages\accelerate\commands\launch.py", line 1159, in launch_command
    multi_gpu_launcher(args)
  File "E:\f5env\Lib\site-packages\accelerate\commands\launch.py", line 793, in multi_gpu_launcher
    distrib_run.run(args)
  File "E:\f5env\Lib\site-packages\torch\distributed\run.py", line 892, in run
    elastic_launch(
  File "E:\f5env\Lib\site-packages\torch\distributed\launcher\api.py", line 133, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\f5env\Lib\site-packages\torch\distributed\launcher\api.py", line 255, in launch_agent
    result = agent.run()
             ^^^^^^^^^^^
  File "E:\f5env\Lib\site-packages\torch\distributed\elastic\metrics\api.py", line 124, in wrapper
    result = f(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^
  File "E:\f5env\Lib\site-packages\torch\distributed\elastic\agent\server\api.py", line 680, in run
    result = self._invoke_run(role)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "E:\f5env\Lib\site-packages\torch\distributed\elastic\agent\server\api.py", line 829, in _invoke_run
    self._initialize_workers(self._worker_group)
  File "E:\f5env\Lib\site-packages\torch\distributed\elastic\metrics\api.py", line 124, in wrapper
    result = f(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^
  File "E:\f5env\Lib\site-packages\torch\distributed\elastic\agent\server\api.py", line 652, in _initialize_workers
    self._rendezvous(worker_group)
  File "E:\f5env\Lib\site-packages\torch\distributed\elastic\metrics\api.py", line 124, in wrapper
    result = f(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^
  File "E:\f5env\Lib\site-packages\torch\distributed\elastic\agent\server\api.py", line 489, in _rendezvous
    rdzv_info = spec.rdzv_handler.next_rendezvous()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\f5env\Lib\site-packages\torch\distributed\elastic\rendezvous\static_tcp_rendezvous.py", line 66, in next_rendezvous
    self._store = TCPStore(  # type: ignore[call-arg]
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: use_libuv was requested but PyTorch was build without libuv support

Can you help me with it? It only happened after I switched to multi-GPU Setup. I am using 2 RTX-4090.

**Reply by SWivid** [reply]
no big idea, haven't encountered that before

maybe try first `accelerate config` as mentioned in readme and also try `train.py`?

**Reply by Anas-KhanWP** [reply]
I Tried but its still going to use_libuv, maybe the reason is that I am on windows and for multi gpu processing nccl is used and nccl is not supported on windows?

**Reply by SWivid** [reply]
dunno
maybe try wsl2?

**Reply by Anas-KhanWP** [reply]
WSL2 didn't show any error, the training started perfectly fine. but still the time taken for it with 2 GPUs is way more then it was with single gpu

**Reply by Anas-KhanWP** [reply]
@lpscr Why batch_size_per_gpu is **3200** when total memory is **24GB** (I explicitly passed the value as 24, just to check) but **1600** when Total Memory is **48GB**? See images Below for reference. Or maybe for 2 GPUs, I need to change this value to **38400 * (Total_gpus)**?

![image](https://github.com/user-attachments/assets/c162f2c4-a6d1-40f4-ae09-a38ef4646189)


With 24GB:
![image](https://github.com/user-attachments/assets/d7c74cfa-97e0-47a8-ac8a-f64f0071f799)

With 48GB:
![image](https://github.com/user-attachments/assets/7e77367e-a07c-4834-a7f1-e24723201f40)
